proposals/0000-scan-takewhile-dropwhile.md

```diff
@@ -0,0 +1,222 @@
+  /// `true` and returning the remainder.
+  func dropWhile(@noescape dropElement: (Self.Generator.Element) throws -> Bool) rethrows -> Self.SubSequence
+  /// Returns a subsequence containing the elements until `takeElement` returns
+  /// `false` and skipping the remainder.
+  func takeWhile(@noescape takeElement: (Self.Generator.Element) throws -> Bool) rethrows -> Self.SubSequence
+}
+```
+
+Also provide default implementations on `SequenceType` that return
+`AnySequence`, and default implementations on `CollectionType` that return a
+slice.
+
+`LazySequenceType` and `LazyCollectionType` will also be extended with
+implementations of `scan(_:combine:)`, `dropWhile(_:)`, and `takeWhile(_:)`
+that return lazy sequence/collection types. Like the lazy `filter(_:)`,
+`dropWhile(_:)` will perform the filtering when `startIndex` is accessed.
+
+Add a global function:
+
+```swift
+/// Returns an infinite sequence of lazy applications of `apply` to the
+/// previous value. For example:
+///
+///     iterate(1, apply: { $0 * 2 }) // yields: 1, 2, 4, 8, 16, 32, 64, ...
+func iterate<T>(initial: T, apply: T -> T) -> IterateSequence<T>
+```
+
+## Detailed design
+
+In addition to the above declarations, provide default implementations based on
+`AnySequence`, similarly to how functions like `dropFirst(_:)` and `prefix(_:)`
+are handled:
+
+```swift
+extension SequenceType {
+  func dropWhile(@noescape dropElement: (Self.Generator.Element) throws -> Bool) rethrows -> AnySequence<Self.Generator.Element>
+  func takeWhile(@noescape takeElement: (Self.Generator.Element) throws -> Bool) rethrows -> AnySequence<Self.Generator.Element>
+}
+```
+
+These default implementations produce an `AnySequence` that wraps an `Array`
+(as the functions must be implemented eagerly to match expected behavior).
+
+Provide default implementations on `CollectionType` as well:
+
+```swift
+extension CollectionType {
+  func dropWhile(@noescape dropElement: (Self.Generator.Element) throws -> Bool) rethrows -> Self.SubSequence
+  func takeWhile(@noescape takeElement: (Self.Generator.Element) throws -> Bool) rethrows -> Self.SubSequence
+}
+```
+
+Extend `LazySequenceType` with lazy versions of the functions:
+
+```swift
+extension LazySequenceType {
+  func scan<T>(initial: T, combine: (T, Self.Generator.Element) -> T) -> LazyScanSequence<Self.Elements, T>
+  func dropWhile(dropElement: (Self.Generator.Element) -> Bool) -> LazyDropWhileSequence<Self.Elements>
+  func takeWhile(takeElement: (Self.Generator.Element) -> Bool) -> LazyTakeWhileSequence<Self.Elements>
+}
+```
+
+The types `LazyScanSequence`, `LazyDropWhileSequence`, and
+`LazyTakeWhileSequence` all conform to `LazySequenceType`.
+
+Extend `LazyCollectionType` with collection variants for `dropWhile(_:)` and
+`takeWhile(_:)` (but not `scan(_:combine:)` because there's no way to recover
+the value from an index without re-scanning from the start):
```

Comment:
I would recommend to proactively make this change from sequence to collection.  The current Dictionary implementation has to keep a reference to the dictionary, this case is not that different.